### PR TITLE
Add an alternate Custom CSS source for 4.7 compat.

### PR DESCRIPTION
**Pull request by georgestephanis**
To Do:

- [x] Populate `/wp-admin/themes.php?page=editcss` with revision viewer -- migrating legacy should be fine.  We can probably move the editor fully into the Customizer instead of backporting that to this page as well.
- [x] Handle data migration from legacy Jetpack CPT to new Core CPT.
- [ ] Code review(s)
- [ ] Ensure all JS and CSS assets have both minified and readable assets (generated by the build process), and properly switch between them.
- [x] Test with a non-`unfiltered_html` user to ensure that Core filters don't mess with CSS syntax thinking it's XSS risk -- escaping carats or the like.

Related:

* https://core.trac.wordpress.org/ticket/35395
* https://core.trac.wordpress.org/ticket/38672
* https://core.trac.wordpress.org/changeset/38829
* #1621 

Core's merged in a Custom CSS provider -- what can we do to play nice with it?

### CODE REVIEW

**Comment by georgestephanis** [comment]
This will fall under the #5539 Master Issue.

**Comment by georgestephanis** [comment]
As of https://github.com/Automattic/jetpack/pull/5527/commits/6bae682be8116106b13a0da00e4bb311bad9a914 Codemirror is working in the Customizer, and if the user doesn't have the `unfiltered_html` cap, it runs the written CSS through CSSTidy.  The cap to `edit_css` is also expanded to anyone with `edit_theme_options`.

File: modules/custom-css/custom-css-4.7.php [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+		return $css;
+	}
+
+	/**
+	 * Override $content_width in customizer previews.
+	 */
+	public static function preview_content_width() {
+		if ( isset( $GLOBALS['wp_customize'] ) ) {
+			if ( isset( $_POST['customized'] ) && isset( $_POST['customize_messenger_channel'] ) ) {
+				$customizations = json_decode( stripslashes( $_POST['customized'] ) );
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: This needs to be replaced with:

```php
$customizations = $GLOBALS['wp_customize']->unsanitized_post_values();
```

Eliminate the `isset( $_POST['customized'] )` check. This will not always be set with changesets.

File: modules/custom-css/custom-css-4.7.php [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+	/**
+	 * Override $content_width in customizer previews.
+	 *
+	 * Runs on `safecss_skip_stylesheet` filter.
+	 */
+	public static function preview_skip_stylesheet( $skip_value ) {
+		if ( isset( $GLOBALS['wp_customize'] ) ) {
+			if ( isset( $_POST['customized'] ) && isset( $_POST['customize_messenger_channel'] ) ) {
+				$customizations = json_decode( stripslashes( $_POST['customized'] ) );
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: This needs to be replaced with:

```php
$customizations = $GLOBALS['wp_customize']->unsanitized_post_values();
```

Eliminate the `isset( $_POST['customized'] )` check. This will not always be set with changesets.

File: modules/custom-css/custom-css-4.7.php [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+		add_action( 'customize_controls_enqueue_scripts', array( __CLASS__, 'customize_controls_enqueue_scripts' ) );
+		add_action( 'customize_register', array( __CLASS__, 'customize_register' ) );
+		add_filter( 'map_meta_cap', array( __CLASS__, 'map_meta_cap' ), 20, 2 );
+
+		// Stuff for stripping out the theme's default stylesheet...
+		add_filter( 'stylesheet_uri', array( __CLASS__, 'style_filter' ) );
+		add_filter( 'safecss_skip_stylesheet', array( __CLASS__, 'preview_skip_stylesheet' ) );
+
+		// Stuff for overriding content width...
+		add_action( 'init', array( __CLASS__, 'preview_content_width' ) );
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: This action, and any others that are supposed to only run in the customizer preview, should only be added in the `customize_preview_init` hook.

File: modules/custom-css/custom-css-4.7.php [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+	/**
+	 * Add Custom CSS section and controls.
+	 */
+	public static function customize_register( $wp_customize ) {
+
+		// SETTINGS
+
+		$wp_customize->add_setting( 'jetpack_custom_css[preprocessor]', array(
+			'default' => '',
+			'transport' => 'postMessage',
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: A `sanitize_callback` can be added here which checks against the `jetpack_custom_css_preprocessors`

File: modules/custom-css/custom-css-4.7.php [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+		) );
+
+		$wp_customize->add_setting( 'jetpack_custom_css[replace]', array(
+			'default' => false,
+			'transport' => 'refresh',
+		) );
+
+		$wp_customize->add_setting( 'jetpack_custom_css[content_width]', array(
+			'default' => '',
+			'transport' => 'refresh',
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: Probably needs a `'sanitize_callback' => array( $wp_customize, 'intval_base10' )`

File: modules/custom-css/custom-css-4.7.php [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+	}
+
+	/**
+	 * Override $content_width in customizer previews.
+	 *
+	 * Runs on `safecss_skip_stylesheet` filter.
+	 */
+	public static function preview_skip_stylesheet( $skip_value ) {
+		if ( isset( $GLOBALS['wp_customize'] ) ) {
+			if ( isset( $_POST['customized'] ) && isset( $_POST['customize_messenger_channel'] ) ) {
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: Checking for `customize_messenger_channel` shouldn't be done, as it may not be present if viewing the customized state in a frontend context.

File: modules/custom-css/custom-css-4.7.php [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+				$customizations = json_decode( stripslashes( $_POST['customized'] ) );
+
+				if ( isset( $customizations->{'jetpack_custom_css[replace]'} ) ) {
+					return $customizations->{'jetpack_custom_css[replace]'};
+				}
+			}
+		}
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: Looking at this again, the logic in this method should be:

```php
global $wp_customize;
if ( ! is_customize_preview() ) {
    return $skip_value;
}
$setting = $wp_customize->get_setting( 'jetpack_custom_css[replace]' );
if ( ! $setting ) {
    return $skip_value;
}
$customized_replace = $setting->post_value();
if ( null !== $customized_replace ) {
    return $customized_replace;
}
return $skip_value;
```

File: modules/custom-css/custom-css-4.7.php [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+				$customizations = json_decode( stripslashes( $_POST['customized'] ) );
+
+				if ( isset( $customizations->{'jetpack_custom_css[content_width]'} ) ) {
+					$width = (int) $customizations->{'jetpack_custom_css[content_width]'};
+
+					if ( $width ) {
+						$GLOBALS['content_width'] = $width;
+					}
+				}
+			}
```
Comment (westonruter) [inline modules/custom-css/custom-css-4.7.php]: As noted below, the logic in this method should be changed to:

```php
global $wp_customize;
if ( ! is_customize_preview() ) {
    return;
}
$setting = $wp_customize->get_setting( 'jetpack_custom_css[content_width]' );
if ( ! $setting ) {
    return;
}
$customized_content_width = $setting->post_value();
if ( null !== $customized_content_width ) {
    $GLOBALS['content_width'] = $customized_content_width;
}
```

Note that the return value from the call to `$setting->post_value()` will have the setting's associated `sanitize_callback` applied.

**Comment by jeherve** [comment]
I gave it a try, but when moving from 4.3.2 to your branch, I lose all my custom CSS. It's not ported over to the Custom CSS editor in the customizer. This happens regardless of whether my existing CSS is preprocessed or not.

I am running `master-stable` + your branch, alongside the most recent version of WordPress `master` (`4.7-beta4-39286` right now).

Am I missing something?

**Comment by georgestephanis** [comment]
@jeherve You're just missing the data migration tool which I'm in  the midst of writing right now.

The new system and old system pulled from totally different post types, so I need to push the old stuff over.

**Comment by georgestephanis** [comment]
@jeherve Migration is written, I've not tested it too closely, but it /should/ work.  I'll test it this evening or tomorrow.  You can run the migration on wp-cli via `wp jetpack custom-css migrate`

It doesn't delete anything, so after migrating, you can  then run `wp jetpack custom-css un-migrate` to let you re-migrate, but each re-migration will increase revisions, as unmigrating just deletes the option that flags it as done, doesn't actually delete the created revisions.

If that makes sense.

Hope it does, if not I'll explain better tomorrow.

File: modules/custom-css/custom-css/js/core-customizer-css-preview.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// Originally based on https://raw.githubusercontent.com/xwp/wp-custom-scss-demo/master/custom-scss-demo-preview.js
+/* globals jpCustomizerCssPreview */
+(function( api, $ ) {
+	if ( api.settingPreviewHandlers ) {
```
Comment (westonruter) [inline modules/custom-css/custom-css/js/core-customizer-css-preview.js]: This will be always empty in WordPress<4.7. Since Jetpack supports WordPress≥4.5, perhaps this entire `core-customizer-css-preview.js` file should be skipped from being enqueued at all, and the `custom_css` setting should always get a `refresh` transport.

File: modules/custom-css/custom-css/js/core-customizer-css.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+		addDesc: function( controlId, desc ) {
+			var control = this._getControl( controlId );
+			if ( control ) {
+				control.append( '<span class="description">' + desc + '<span>' );
+			}
+		},
+		/**
+		 * Helper function to qet a control by ID
+		 * @param  {string} controlId Control ID
+		 * @return {object}           Customizer Control object
```
Comment (westonruter) [inline modules/custom-css/custom-css/js/core-customizer-css.js]: This is returning a jQuery object not an instance of `wp.customize.Control`

File: modules/custom-css/custom-css/js/core-customizer-css.js [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+			}, this ) );
+
+			// should we use CodeMirror?
+			if ( this.opts.useRichEditor ) {
+				this.initCodeMirror();
+			} else {
+				this.$input.removeClass( 'hidden' );
+			}
+
+			api.bind( 'ready', _.bind( this.addLabels, this ) );
```
Comment (westonruter) [inline modules/custom-css/custom-css/js/core-customizer-css.js]: It shouldn't need to bind to the global `ready` here since this is the control's own `ready` method. Can't it just do `this.addLabels()`?

File: modules/custom-css/migrate-to-core.php [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+				$pre .= $jetpack_css_revision->post_content_filtered;
+
+				$css .= "\r\n\r\n/*\r\n\t" . esc_js( __( 'CSS Migrated from Jetpack:', 'jetpack' ) ) . "\r\n*/\r\n\r\n";
+				$css .= call_user_func( $preprocessors[ $preprocessor ]['callback'], $jetpack_css_revision->post_content_filtered );
+			} else {
+				$css .= "\r\n\r\n/*\r\n\t" . esc_js( __( 'CSS Migrated from Jetpack:', 'jetpack' ) ) . "\r\n*/\r\n\r\n";
+				$css .= $jetpack_css_revision->post_content;
+			}
+
+			wp_update_post( array(
```
Comment (westonruter) [inline modules/custom-css/migrate-to-core.php]: The `array` needs to be wrapped in `wp_slash()` since `wp_update_post()` expects (aggravatingly) its input to be pre-slashed. So if `wp_slash()` isn't added, then any slashes in `$css` or `$pre` will get dropped and a CSS parse error could result.

File: modules/custom-css/custom-css/js/core-customizer-css-preview.js (hunk 9 above)
Comment (georgestephanis) [inline modules/custom-css/custom-css/js/core-customizer-css-preview.js]: I'm not expecting much difficulty here, as this will only be enqueued if the new class is pulled in by the [`if ( function_exists( 'wp_get_custom_css' ) )` ](https://github.com/Automattic/jetpack/blob/80d4b437bcc6f87486394e63de308885053a26fc/modules/custom-css.php#L18)conditional.  I just left it there for anyone running trunk in an intermediate changeset.

File: modules/custom-css/custom-css/js/core-customizer-css.js (hunk 11 above)
Comment (georgestephanis) [inline modules/custom-css/custom-css/js/core-customizer-css.js]: I honestly don't know, this is from the existing wpcom custom-css-in-customizer code.  Will check out who committed it originally and ask them to confirm.

File: modules/custom-css/custom-css/js/core-customizer-css.js (hunk 10 above)
Comment (georgestephanis) [inline modules/custom-css/custom-css/js/core-customizer-css.js]: Yup, looks like this was a typo in the original wpcom 80185 changeset.

File: modules/custom-css/custom-css/js/core-customizer-css.js (hunk 11 above)
Comment (mattwiebe) [inline modules/custom-css/custom-css/js/core-customizer-css.js]: I have a vague recollection of needing this action delayed, but I wrote this code a long time ago. Swapping to `this.addLabels()` would be fine if it works.

**Comment by samhotchkiss** [comment]
Tried this out, and no dice :(

red-bg-go-byebye

What I did:
• In 4.6.1, with theme "Twenty Fifteen", I set the background color to red using Jetpack custom CSS
• Upgraded to 4.7

Problems:
• My CSS changes went away
• My CSS changes also disappeared from the Edit CSS page
• There are now two Edit CSS links in the WP admin menu

**Comment by georgestephanis** [comment]
The two edit css links are expected.  The one with the lightning bolt goes directly to the customizer, the one without goes to our now-sparse legacy page.  It's a convenience for me while I'm building things and deciding what we should use in the end.

I'm not completely certain why the changes aren't migrating, let me purge all my existing db stuff and start over to see what I can come up with.

**Comment by georgestephanis** [comment]
@samhotchkiss Migration should work properly now, there seem to have been some issues with when there were no revisions, but a post saved -- I was only checking revisions, not the post directly.

**Comment by samhotchkiss** [comment]
Hmmm, now, after upgrading, my custom CSS is the ONLY css:

https://cloudup.com/c05iYFwUGXg

File: modules/custom-css.php [hunk 13]
```diff
@@ -13,6 +13,28 @@
  */
 
 function jetpack_load_custom_css() {
+	// If WordPress has the core version of Custom CSS, load our new version.
+	// @see https://core.trac.wordpress.org/changeset/38829
+	if ( function_exists( 'wp_get_custom_css' ) ) {
+		if ( ! function_exists( 'wp_update_custom_css_post' ) ) {
+			wp_die( 'Please run a SVN up to get the latest version of trunk, or update to at least 4.7 RC1' );
```
Comment (jeherve) [inline modules/custom-css.php]: I assume we'll do something else until Jetpack requires 4.7?

File: modules/custom-css/custom-css-4.7.php [hunk 14]
```diff
@@ -0,0 +1,781 @@
...
+		}
+		return $caps;
+	}
+
+	public static function admin_menu() {
+		// Add in our legacy page to support old bookmarks and such.
+		add_theme_page( __( 'CSS', 'jetpack' ), __( 'Edit CSS', 'jetpack' ), 'edit_theme_options', 'editcss', array( __CLASS__, 'admin_page' ) );
+
+		// Add in our new page slug that will redirect to the customizer.
+		$hook = add_theme_page( __( 'CSS', 'jetpack' ), __( 'Edit CSS ⚡', 'jetpack' ), 'edit_theme_options', 'editcss-customizer-redirect', array( __CLASS__, 'admin_page' ) );
```
Comment (jeherve) [inline modules/custom-css/custom-css-4.7.php]: Will we eventually only have the regular "Edit CSS" menu item (without the ⚡ emoji), and have it point to `editcss-customizer-redirect`?

File: modules/custom-css/custom-css-4.7.php [hunk 15]
```diff
@@ -0,0 +1,781 @@
...
+
+		<ul>
+		<?php foreach ( $themes as $theme_stylesheet => $data ) :
+			if ( $stylesheet === $theme_stylesheet ) {
+				continue;
+			}
+			$revisions = wp_get_post_revisions( $data['post']->ID, array( 'posts_per_page' => 1 ) );
+			$revision = array_shift( $revisions );
+			?>
+			<li><a href="<?php echo esc_url( get_edit_post_link( $revision->ID ) ); ?>"><?php echo esc_html( $data['label'] ); ?></a>
```
Comment (jeherve) [inline modules/custom-css/custom-css-4.7.php]: This triggers a PHP notice for me:

```
PHP Notice:  Trying to get property of non-object in /wp-content/plugins/jetpack/modules/custom-css/custom-css-4.7.php on line 179
```

**Comment by samhotchkiss** [comment]
Following up on the issue that @jeherve and I noticed-- it looks like this option is getting selected by default:

image

Should be easy enough to remedy, hopefully!

File: modules/custom-css/custom-css-4.7.php (hunk 15 above)
Comment (samhotchkiss) [inline modules/custom-css/custom-css-4.7.php]: Yep, I experienced that , too

File: modules/custom-css/custom-css-4.7.php (hunk 14 above)
Comment (samhotchkiss) [inline modules/custom-css/custom-css-4.7.php]: Yes, that's correct (I discussed this with George, who confirmed)

File: modules/custom-css.php (hunk 13 above)
Comment (samhotchkiss) [inline modules/custom-css.php]: This shouldn't be necessary (we'll drop this logic before shipping), it's just to prevent people from running this branch against an earlier 4.7 beta than RC1

File: _inc/client/search/index.jsx [hunk 16]
```diff
@@ -105,6 +105,7 @@ export const SearchResults = ( {
 		let isPro = 'scan' === element[0]
 				 || 'akismet' === element[0]
 				 || 'backups' === element[0]
+				 || 'videopress' === element[0]
```
Comment (zinigor) [inline _inc/client/search/index.jsx]: This needs to be removed as a leftover from moving stuff from `master`.

File: json-endpoints/jetpack/class.jetpack-json-api-themes-modify-endpoint.php [hunk 17]
```diff
@@ -64,4 +64,48 @@ function update() {
 		return true;
 	}
 
+	function update_translations() {
```
Comment (zinigor) [inline json-endpoints/jetpack/class.jetpack-json-api-themes-modify-endpoint.php]: I'm assuming the translations and endpoint stuff shouldn't be a part of this PR, right?

File: modules/custom-css/custom-css/js/core-customizer-css-preview.js [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+			}
+
+			// No special providers, just write what we got.
+			deferred = new $.Deferred();
+			setting = api( 'custom_css[' + api.settings.theme.stylesheet + ']' );
+			_.each( partial.placements(), function( placement ) {
+				placement.container.text( setting.get() );
+			} );
+
+			deferred.resolve();
```
Comment (zinigor) [inline modules/custom-css/custom-css/js/core-customizer-css-preview.js]: This is kind of confusing, can we just `return $.Deferred().resolve().promise();` and leave a comment that we return a dummy promise?

File: modules/videopress/shortcode.php [hunk 19]
```diff
@@ -39,7 +39,7 @@ function videopress_shortcode_callback( $attr ) {
 	$defaults = array(
 		'w'               => 0,     // Width of the video player, in pixels
 		'at'              => 0,     // How many seconds in to initially seek to
-		'hd'              => false, // Whether to display a high definition version
+		'hd'              => true,  // Whether to display a high definition version
```
Comment (zinigor) [inline modules/videopress/shortcode.php]: Needs to be reverted to branch-4.4 state.

File: sync/class.jetpack-sync-module-comments.php [hunk 20]
```diff
@@ -32,6 +32,10 @@ public function init_listeners( $callable ) {
 				add_action( $comment_action_name, $callable, 10, 2 );
 			}
 		}
+
```
Comment (zinigor) [inline sync/class.jetpack-sync-module-comments.php]: Sync-related changes shouldn't be here either.

File: modules/custom-css/custom-css/js/core-customizer-css-preview.js (hunk 18 above)
Comment (georgestephanis) [inline modules/custom-css/custom-css/js/core-customizer-css-preview.js]: This is a copy/paste of @westonruter's code from https://github.com/xwp/wp-custom-scss-demo/blob/master/custom-scss-demo-preview.js -- I'm not familiar enough with promises to have a position to it.

File: json-endpoints/jetpack/class.jetpack-json-api-themes-modify-endpoint.php (hunk 17 above)
Comment (georgestephanis) [inline json-endpoints/jetpack/class.jetpack-json-api-themes-modify-endpoint.php]: Correct, this is from when I updated the PR from master, but then Sam rebased it to the older branch.  Awkward cruft.  I'm thinking it'll probably get smushed with everything else trimmed off before it's merged in.

File: _inc/client/search/index.jsx (hunk 16 above)
Comment (georgestephanis) [inline _inc/client/search/index.jsx]: Cruft from master having been merged in before this was rebased onto branch-4.4

Honestly, @samhotchkiss -- does it have to be rebased onto branch-4.4?  Couldn't it just be rebased onto Master, but merged into 4.4 and master both when approved?  Or smush-merged into master, then cherry-picked across?

**Comment by georgestephanis** [comment]
@samhotchkiss @jeherve I think the problem is the migration script -- the option got changed from whether it's add-on, to whether it should replace, but the setting was never flipped.

https://github.com/Automattic/jetpack/pull/5527/files#diff-2db59513845c02c1b1615893ff5f351fR196

**Comment by georgestephanis** [comment]
@jeherve I think I've got all your concerns now -- but btw, the revision list you see is just a list of themes that have revisions, not individual revisions.  New CPT stores each theme in its own post, so switching back to a theme restores its prior customizations.

**Comment by jeherve** [comment]
I gave it another try.

- The Revisions link now includes a reference, but that page doesn't load the Revisions screen, it just reloads the same page:
screen shot 2016-11-30 at 12 33 27
- The `jetpack_custom_css[replace]` Customizer option still appears to be reversed: to see the original theme's CSS, you have to uncheck the box.
- My previous CSS doesn't get migrated.
- New CSS appears to be saved properly, but when going back to `themes.php?page=editcss`, the last revisions data isn't updated.

**Comment by folletto** [comment]
I added two notes in terms of design review here: https://github.com/Automattic/jetpack/issues/5539#issuecomment-263866150

tl;dr:

* Removed padding for the editor to maximize the area and align with future design
* Moved the revision/help links at the bottom, to align with future design

The comments there also contain rough CSS to make it work.

File: _inc/client/search/index.jsx (hunk 16 above)
Comment (zinigor) [inline _inc/client/search/index.jsx]: Sam has a good point that it's easier to test it if we have it based on `branch-4.4`. We'll squash the end result and cherry-pick it over to `master`.

**Comment by zinigor** [comment]
I have just rebased the branch, please reset your local copies if you were working on this. Be careful, do not `git pull` because it will merge the new branch into the old branch.

**Comment by samhotchkiss** [comment]
@georgestephanis -- looks like the "Start Fresh" box is still getting checked

**Comment by georgestephanis** [comment]
Closing this PR in lieu of one I'm about to make.

**Comment by georgestephanis** [comment]
Whoops, it was fixed by @zinigor !

**Comment by georgestephanis** [comment]
@samhotchkiss So the only thing left to go is adding an ability to 'view' the css of an inactive theme when that theme has just the content saved, but no revisions.  I'm working on it, and we can add that in a beta2 if you like.

This is totally runnable and testable now as a beta 1.

**Comment by georgestephanis** [comment]
(alternately, I'll have the viewer done in about 20m)

**Comment by georgestephanis** [comment]
Okay.  That should do it.

:shipit:

**Comment by georgestephanis** [comment]
Can someone get a last code review on this?

**Comment by folletto** [comment]
Any chance you could push in the design changes from #5539 before or right after this PR gets merged?

Changes:

1. Attach the bar and introduce hover/focus styles for accessibility and alignment (CSS provided).
2. Change label to "See full history".
3. Change label to "Help".

Full details, CSS to grab and screenshot [here](https://github.com/Automattic/jetpack/issues/5539#issuecomment-264202160).
